src/models/predicate/mod.rs

```diff
@@ -0,0 +1,95 @@
+use crate::{Error, Result};
+use std::{
+    collections::BTreeMap,
+    fmt::{Display, Formatter, Result as FmtResult},
+};
+
+#[derive(Debug, Serialize, Deserialize, Hash, PartialEq, EnumIter, Clone, Copy)]
+pub enum PredicateVersion {
+    NonExist,
+    // V0_2,
+}
+
+impl Display for PredicateVersion {
+    fn fmt(&self, fmt: &mut Formatter) -> FmtResult {
+        match self {
+            // PredicateVersion::V0_2 => fmt.write_str("https://in-toto.io/Link/v0.2")?,
+            PredicateVersion::NonExist => fmt.write_str("nonexist")?,
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub enum PredicateWrapper {
+    // V0_2(PredicateV02),
+    NonExist(PredicateNonexist),
+}
+
+impl PredicateWrapper {
+    /// Convert from enum `PredicateWrapper` to trait `PredicateLayout`
+    pub fn into_trait(self) -> Box<dyn PredicateLayout> {
+        match self {
+            PredicateWrapper::NonExist(link) => Box::new(link),
+            // PredicateWrapper::V0_2(link) => Box::new(link),
+        }
+    }
+
+    /// Construct `PredicateWrapper` from MetaData
+    pub fn from_meta(meta: LinkMetadataBuilder, version: PredicateVersion) -> Self {
+        match version {
+            PredicateVersion::NonExist => Self::NonExist(PredicateNonexist::from(meta)),
+            // PredicateVersion::V0_2 => Self::V0_2(PredicateV02::from(meta)),
+        }
+    }
+
+    /// Standard deserialize for PredicateWrapper by its version
+    pub fn from_bytes(bytes: Vec<u8>, version: PredicateVersion) -> Result<Self> {
+        match version {
+            PredicateVersion::NonExist => serde_json::from_slice(&bytes)
+                .map(Self::NonExist)
+                .map_err(|e| e.into()),
+            // PredicateVersion::V0_2 => serde_json::from_slice(&bytes)
+            //     .map(Self::V0_2)
+            //     .map_err(|e| e.into()),
+        }
+    }
+    /// Auto deserialize for PredicateWrapper by any possiable version.
+    pub fn try_from_bytes(bytes: Vec<u8>) -> Result<Self> {
+        let mut link: Result<PredicateWrapper> =
+            Err(Error::Programming("no available bytes parser".to_string()));
+        for version in PredicateVersion::iter() {
+            link = PredicateWrapper::from_bytes(bytes.clone(), version);
+            if link.is_ok() {
+                break;
+            }
+        }
+        link
+    }
+}
+
+pub trait PredicateLayout {
+    /// The version of predicate
+    fn version(&self) -> PredicateVersion;
+    /// Convert from trait `PredicateLayout` to enum `PredicateWrapper`
+    fn into_enum(self: Box<Self>) -> PredicateWrapper;
+    /// Standard serialize for PredicateLayout
+    fn to_bytes(&self) -> Result<Vec<u8>>;
+    fn name(&self) -> Option<&String>;
+    fn materials(&self) -> Option<&BTreeMap<VirtualTargetPath, TargetDescription>>;
+    fn env(&self) -> Option<&BTreeMap<String, String>>;
+    fn byproducts(&self) -> Option<&BTreeMap<String, String>>;
```

Comment:
Hi, do we need return `Option<&ByProducts>` here for https://github.com/in-toto/in-toto-rs/commit/25493433cdcf74f9031207269f5ffb3288d6fed8 ?